### PR TITLE
add *(::Union{Regex, AbstractString, AbstractChar}...)

### PR DESCRIPTION
**Pull request by rfourquet**
I needed it recently and got frustrated that it doesn't work:
```
julia> const width = 7
julia> replace(line, r"^" * ' '^width, "") # ok, doesn't work, have to look up how to do repetition in PCRE...
julia> replace(line, r"^ {$width}", "") # ah crap, cannot do string interpolation...
julia> replace(line, r"^ {7}, "") # NOTE:  this 7 corresponds to `width`, don't forget to update it when needed
```
So yes there is a way to achieve what is needed with the `Regex` constructor, but allowing `*` is more regex-noobie friendly.

I don't know if it's a good idea, but don't foresee a problem with it; note that I didn't go so far as making `Regex <: AbstractString`, as I didn't expect to get support for that ;-)

### CODE REVIEW

**Comment by waldyrious** [comment]
> allowing `*` is more noobie friendly.

I assume you mean "regex noobie", because `r"^" * ' '^7` looks almost like obfuscated code to me. IMHO the second option (`r"^ {$width}"`) is the most readable, if that's a possibility.

**Comment by rfourquet** [comment]
Yes you are right, I edited to "regex-noobie friendly" ! For the option you find more readble, it would require more work, and as `$` has a special meaning in regexes it may be subtle to implement.

**Comment by StefanKarpinski** [comment]
Related: regular languages with `*` as concatenation of patterns and `+` as alternation of patterns (i.e. `|` in regex syntax) form a semiring. In fact this is the original [Kleene algebra](https://en.wikipedia.org/wiki/Kleene_algebra) – i.e. a semiring with the additional requirement that `+` is idempotent (`a + a = a` for all `a`). This naturally extends to classes of equivalent regular expressions and strings are embedded in this space as wildcard-free patterns (i.e. the string as a literal pattern).

In other words, extending `*` to work on regular expressions mixtures of regexes and strings is a completely natural extension of how `*` works on strings already. It would also make sense to define `+` on strings and regexes to compute the alternation of its arguments. That said, I'm a bit reluctant to start adding this kind of thing at this point since once we start doing this, I feel like we should do it thoroughly – i.e. implement the entire Kleene algebra. There would definitely be some people who would not be thrilled about `"foo" + "bar"` producing `re"(foo|bar)"` 😬.

**Comment by rfourquet** [comment]
[Someone else](https://github.com/JuliaLang/julia/issues/11030#issuecomment-326772422) had a similar observation: 

> In linear algebra, there is a unary operator, the adjoint operator, often denoted by *. In Julia as well as Matlab, the adjoint operator is give by a single quote ('), since * is generally used for multplication. So I would propose string operators in Julia to be (\*,+,') for concatenation, alternation, and Kleene star respectively.

In this PR I clearly don't intend to go beyond multiplication:  is it still too early to decide to do that, or shall I add tests and documentation?

**Comment by StefanKarpinski** [comment]
Go for it!

**Comment by rfourquet** [comment]
Freshely rebased and documented, ready for reviews :)

File: base/regex.jl
Comment (StefanKarpinski) [inline base/regex.jl]: This could be supported with [internal option setting](https://www.pcre.org/original/doc/html/pcrepattern.html#SEC13), probably best combined with a [non-capturing subpattern](https://www.pcre.org/original/doc/html/pcrepattern.html#SEC14). So something like this:
```jl
"(?sm:$(r1.pattern))(?i:$(r2.pattern))"
```
This could work uniformly but it would be nice to avoid the extra line noise in the presumably common case where the options match. You could get even fancier and keep options that are shared by all of the regexes in the result regex's global options while only putting the options that are specific to each subpattern in it's option set.

**Review by StefanKarpinski — CHANGES_REQUESTED** [review]
My immediate instinct is to worry about composing expressions via string operations. And indeed, I think there are some quite broken cases here. E.g. `r"this|that" * r"cat|hat"` would result in `r"this|thatcat|hat"` which is not the correct juxtaposition at all. Similarly, for repetition `r"this|that"^2` results in `r"this|thatthis|that"` which is also wrong. I think that if you unconditionally wrap regex patterns in non-capturing subpatterns that this might be correct. So you'd produce `r"(?:this|that)(?:cat|hat)"` for the first example. For repetition, I think it's wasteful to actually textually repeat the regex pattern since regular expressions already support repetition as a language feature. So for the second case, I would instead generate `r"(?:this|that)^{2}"`.

Other features that we may want to consider while thinking about such an API expansion are:
* alternation of regexes and even strings, producing regexes
* repetition by a range of indices to get between min and max number of repetitions
* an operator/function for Kleene star?

For example, we could use `+` for alternation since that makes `*` and `+` a semiring on regexes with strings embedded in them as "simple" patterns. We could also use something like `p^(a:b)` to generate `r"(?:$p){$a,$b}"` if you follow my pseudo-syntax here.

**Comment by StefanKarpinski** [comment]
So overall todos here are:

* wrap every pattern in non-capturing groups `(?$opts:$pattern)` before doing operations
* for concatenation, collect shared options in the resulting regex's global options
* put non-shared options in the `opts` part of each components' non-capturing group
* use `{n}` notation for repetition of regexes

**Comment by rfourquet** [comment]
Thanks so much for taking the time to set out the logic! and for linking to documentation. I hadn't even started to think about this sophistication, and didn't know the PCRE features you indicated to implement the logic.
(I may have a invented a new way to learn PCRE: make a PR touching regexes and wait for feedback :wink:)

So I implemented this, but with the limitation that only the four options corresponding to "imsx" can be different between the multiplied regexes (IIUC, these are the only ones which can be indicated "internally").

Concerning the extension of the API, I unfortunately have nothing to contribute to your ideas, but it make sense to me and I would be happy to implement them. For a Kleene star operator, it is unfortunate that `*` is already taken (and similarly for the "one or more" operation and  `+`).

File: base/regex.jl
Comment (rfourquet) [inline base/regex.jl]: Oups, forgot to update the docstrings, will do next time I push. Does it need more explanations? ("concatenate regexes" is quite minimal).

File: test/regex.jl
Comment (StefanKarpinski) [inline test/regex.jl]: Should this really work like this? I would expect the string `"b"` to only match itself?

File: base/regex.jl
Comment (StefanKarpinski) [inline base/regex.jl]: So I think this is not how we want this to work: I think that a string or character should be converted to a regex that only matches that exact string or character, not to a regex with that string or character as its content. Also, I think there's a case to be made for making this a method of convert. For a clearer example where it matters, should `r"a|b" * "c|d"` produce `r"(?:a|b)(?:c|d)"` which would match `ac`, `ad`, `bc` and `bd` or should it produce something like `r"(?:a|b)(?:c\|d)"` which would match `ac|d` and `bc|d`? I think it has to be the latter.

**Review by StefanKarpinski — APPROVED** [review]
The regex composition stuff is looking good now but I think the string embedding into regular expressions is still wrong: strings and characters need to be embedded as patterns which match only that literal string or character.

**Review by StefanKarpinski — CHANGES_REQUESTED** [review]
Meant to submit that review as "request changes" but picked the wrong radio button.

**Comment by StefanKarpinski** [comment]
In fully generality, this requires being able to correctly regex-quote strings, which is definitely doable, but not entirely straightforward, I suspect. It may be easier to just skip that part of this change and keep it focused on the pure regex bit.

**Comment by rfourquet** [comment]
> skip that part of this change and keep it focused on the pure regex bit.

So allow only multiplication of regexes, not regexes + strings + chars?

File: base/regex.jl
Comment (rfourquet) [reply base/regex.jl]: OK, there seems to be a [solution](https://www.pcre.org/original/doc/html/pcrepattern.html#SEC5) after all: enclose the string between "\Q" and "\E". Remains to decide which behavior is the best, I don't have a strong opinion. But it's true that if one wants the regex interpretation of characters, it's easy enough to wrap the string in a call to `Regex`.

**Comment by StefanKarpinski** [comment]
You can certainly include that too, but does it makes sense to you that it should treat strings as a regex that only matches that string? Of you’re cool with implementing that, I have no objection.

**Comment by rfourquet** [comment]
I don't know if you saw my [comment](https://github.com/JuliaLang/julia/pull/23422#discussion_r271817950) above, but there seems to be a trivial solution, using quoting between "\Q" and "\E". So I updated the PR with this.

> does it makes sense to you that it should treat strings as a regex that only matches that string?

It sure makes sense, but as I said I don't have a strong opinion and anyway it (my opinion) shouldn't matter as I didn't do much regexing in my life. Your call!

File: base/regex.jl
Comment (StefanKarpinski) [inline base/regex.jl]: Ok, you're going to want to kill me but what if `s` contains `\E`?

**Review by StefanKarpinski — CHANGES_REQUESTED** [review]
This is getting there. I think if you just handle `\E` in strings we should be good.

File: base/regex.jl
Comment (rfourquet) [reply base/regex.jl]: Hahaha! actually, I should have known better with my background in cybersecurity (cf. injection attacks) ;-)

**Review by StefanKarpinski — APPROVED** [review]

**Comment by StefanKarpinski** [comment]
This looks good to me now. We’re halfway to having a fully algebra on regular expressions!

**Comment by rfourquet** [comment]
So the intersection of the failing tests between this CI run and the previous one is empty (previous failures: [travis](https://travis-ci.org/JuliaLang/julia/builds/524491711?utm_source=github_status&utm_medium=notification), [tester_macos64](https://build.julialang.org/#/builders/69/builds/988), [tester_linux32](https://build.julialang.org/#/builders/13/builds/1957)), so I will assume they are not related.

I will merge soon if no objection.

**Comment by StefanKarpinski** [comment]
Go for it! Squash or not as appropriate.

File: NEWS.md
Comment (fredrikekre) [inline NEWS.md]: Bad rebase?

Comment (StefanKarpinski) [reply NEWS.md]: Good catch. Fixed in https://github.com/JuliaLang/julia/pull/31874

Comment (rfourquet) [reply NEWS.md]: Ooops, sorry about that!

**Pull request merged**
